packages/std/src/memory.rs

```diff
@@ -116,55 +54,167 @@ unsafe impl RegionSource for Vec<u8> {
     }
 }
 
-unsafe impl<T: ?Sized> RegionSource for &T
+mod sealed {
+    pub trait Sealed: 'static {}
+
+    impl Sealed for super::Owned {}
+
+    impl Sealed for super::Borrowed {}
+}
+
+pub trait Ownership: sealed::Sealed + 'static {}
+
+impl<T> Ownership for T where T: sealed::Sealed {}
+
+pub struct Owned {}
+
+pub struct Borrowed {}
+
+/// Describes some data allocated in Wasm's linear memory.
+/// A pointer to an instance of this can be returned over FFI boundaries.
+///
+/// This struct is crate internal since the cosmwasm-vm defines the same type independently.
+#[repr(C)]
+pub struct Region<T: Ownership> {
+    /// The beginning of the region expressed as bytes from the beginning of the linear memory
+    pub offset: u32,
+    /// The number of bytes available in this region
+    pub capacity: u32,
+    /// The number of bytes used in this region
+    pub length: u32,
+
+    _marker: PhantomData<T>,
+}
```

Comment:
This type in std could use a size check as well (for all common T)

Reply:
Not sure if we need a size check for that. The layout of `PhantomData` is guaranteed to be size 0, align 1 (ref: <https://doc.rust-lang.org/stable/core/marker/struct.PhantomData.html#layout-1>)

But we can definitely add one